### PR TITLE
Fix py3 issues

### PR DESCRIPTION
**Pull request by heplesser**
This PR fixes a number of Python 3 issues, including incorrect imports in pynest/nest/lib, thus addressing #617 among other issues. It is a superset of #618.

### CODE REVIEW

File: pynest/nest/lib/hl_api_connections.py [hunk 1]
```diff
@@ -453,15 +452,15 @@ def _is_subnet_instance(gids):
...
-        # Turn off _deprecation_warning as users shouldn't change
-        # implementation of this function, it is done by the developers
-        deprecation_bool = hlh._deprecation_warning['GetChildren']
-        hlh._deprecation_warning['GetChildren'] = False
+        # Turn off deprecation warning to avoid confusing users with
+        # internals.
+        # deprecation_status = suppress_deprecation_warning('GetChildren')
 
-        GetChildren(gids)
+        with SuppressedDeprecationWarning(['GetChildren']):
```
Comment (Silmathoron) [inline pynest/nest/lib/hl_api_connections.py]: Where is this defined? Couldn't find it in ``hl_api_helper``
**EDIT:** ok, just below ^^

File: pynest/nest/lib/hl_api_connections.py [hunk 2]
```diff
@@ -453,15 +452,15 @@ def _is_subnet_instance(gids):
...
+            GetChildren(gids)
 
         # Need to reset the deprecation warning to its old value
-        hlh._deprecation_warning['GetChildren'] = deprecation_bool
+        # restore_deprecation_warning('GetChildren', deprecation_status)
```
Comment (Silmathoron) [inline pynest/nest/lib/hl_api_connections.py]: why is this here if it is commented (same question for L457)

File: pynest/nest/lib/hl_api_helper.py [hunk 3]
```diff
@@ -424,63 +454,43 @@ def model_deprecation_warning(model):
         warnings.warn('\n' + text)
 
 
-def turn_off_deprecation_warning(deprecated_func_name):
-    """Turns off deprecation warnings on SLI and Python level for given
-    function.
+class SuppressedDeprecationWarning(object):
```
Comment (Silmathoron) [inline pynest/nest/lib/hl_api_helper.py]: I think this should not be exposed to users, so maybe add a ``_`` before and a ``__all__`` in the file?

File: pynest/nest/lib/hl_api_helper.py [hunk 4]
```diff
@@ -402,6 +400,38 @@ def broadcast(item, length, allowed_types, name="item"):
     return item
 
 
+@check_stack
```
Comment (Silmathoron) [inline pynest/nest/lib/hl_api_helper.py]: Why did you switch it here? (I don't mind, though, just curious)

File: pynest/nest/lib/hl_api_subnets.py [hunk 5]
```diff
@@ -49,16 +47,9 @@ def PrintNetwork(depth=1, subnet=None):
...
-        # developers.
-        deprecation_bool, verbosity_level = turn_off_deprecation_warning(
-            'CurrentSubnet')
-        subnet = CurrentSubnet()
-        # Need to reset the deprecation warning to its old value
-        turn_on_deprecation_warning('CurrentSubnet', deprecation_bool,
-                                    verbosity_level)
-
+        # Avoid confusing user by deprecation warning
+        with SuppressedDeprecationWarning('CurrentSubnet'):
```
Comment (Silmathoron) [inline pynest/nest/lib/hl_api_subnets.py]: Note that if you indeed chose to make this not available to users (see comment above), you'll need to import it explicitly

File: topology/pynest/hl_api.py [hunk 6]
```diff
@@ -1619,20 +1619,12 @@ def GetTargetNodes(sources, tgt_layer, tgt_model=None, syn_model=None):
...
-
-    # obtain local nodes in target layer, to pass to GetConnections
-    tgt_nodes = nest.GetLeaves(tgt_layer,
-                               properties={
-                                   'model': tgt_model} if tgt_model else None,
-                               local_only=True)[0]
-    # Need to reset the deprecation warning to its old value
-    nest.turn_on_deprecation_warning('GetLeaves', deprecation_bool,
-                                     verbosity_level)
+    with nest.SuppressDeprecationWarning('GetLeaves'):
```
Comment (Silmathoron) [inline topology/pynest/hl_api.py]: Same comment as above on accessibility of this function

File: examples/nest/Potjans_2014/spike_analysis.py [hunk 7]
```diff
@@ -57,16 +57,16 @@
...
+print(gids)
+print()
 
 # number of populations
 
 num_pops = len(gids)
-print 'Number of populations:'
-print num_pops
-print
+print('Number of populations:')
```
Comment (heplesser) [inline examples/nest/Potjans_2014/spike_analysis.py]: I agree with you in principle, but in this case I prioritised getting things done over doing them perfectly. Example beautification can come in a separate PR later.

File: extras/help_generator/writers.py [hunk 8]
```diff
@@ -59,10 +59,10 @@ def write_help_html(doc_dic, helpdir, fname, sli_command_list, keywords):
     hlplist = []
     name = ''
 
-    for key, value in doc_dic.iteritems():
+    for key, value in doc_dic.items():
```
Comment (heplesser) [inline extras/help_generator/writers.py]: I think wrapping with `iter()` will not provide any gains and rather make the code less readable. The only point of using `iteritems()` in Python 2 is to get a generator instead of a full list of items. `iter(d.items())` will still create the full list and then wrap an iterator around it in Python 2.  And `for k, v in d.items():` is the pythonic way as far as I know.

File: pynest/nest/lib/hl_api_connections.py (hunk 2 above)
Comment (heplesser) [inline pynest/nest/lib/hl_api_connections.py]: Fixed.

File: pynest/nest/lib/hl_api_helper.py (hunk 4 above)
Comment (heplesser) [inline pynest/nest/lib/hl_api_helper.py]: I ran into a circular inclusion problem otherwise. The `SuppressedDeprecationWarning` class definition fits best here among the helpers. It needs to manipulate the verbosity level. Having the verbosity functions defined in `hl_api_info.py` requires helper to include info an vice-versa. Moving the verbosity function definitions here means that `hl_api_helper.py` does not need to include any other `hl_api_*.py` files.

File: examples/nest/Potjans_2014/spike_analysis.py (hunk 7 above)
Comment (Silmathoron) [inline examples/nest/Potjans_2014/spike_analysis.py]: No problem

File: extras/help_generator/writers.py (hunk 8 above)
Comment (Silmathoron) [inline extras/help_generator/writers.py]: Ok, I thought they would find a clever way to do this in py2... but if the list is created, then there is indeed really no point

File: pynest/nest/lib/hl_api_helper.py (hunk 4 above)
Comment (Silmathoron) [inline pynest/nest/lib/hl_api_helper.py]: ok

File: pynest/nest/lib/hl_api_helper.py (hunk 3 above)
Comment (heplesser) [inline pynest/nest/lib/hl_api_helper.py]: Fixed. I will add explicit imports where needed.

Comment (heplesser) [inline pynest/nest/lib/hl_api_helper.py]: I thought about it a little more. `SuppressedDeprecationWarning` is also needed in the Topology hl_api, so I would have to access it there as a private member; possible, but slightly breaking with the logic of privacy. It is also not an implementation detail that may change at any time and does not require privacy from that perspective. Users should realise that they should use this context with care based on its name and documentation. Thus, I will leave the name as it is after all.

Comment (Silmathoron) [inline pynest/nest/lib/hl_api_helper.py]: Fair enough
